src/Zigzag_persistence/include/gudhi/zigzag_persistence.h

```diff
@@ -0,0 +1,458 @@
+
+   private:
+    birth_dictionary birthToPos_; /**< birth_to_pos_[i] < birth_to_pos_[j] iff i <b j */
+    index maxBirthPos_;            /**< is strictly larger than any other birth so far */
+    index minBirthPos_;            /**< is strictly smaller than any other birth so far */
+  };
+
+ public:
+  /**
+   * @brief Constructor of the Zigzag_persistence class.
+   * @details After construction of the class, the zigzag filtration should be given in a streaming like way, i.e.,
+   * call @ref insert_face, @ref remove_face or @ref apply_identity for each step of the filtration in order of
+   * the filtration. The pairs of birth and death indices are retrieved via the given callback method every time
+   * a pair is closed. To retrieve the open pairs (corresponding to infinite bars),
+   * use @ref get_current_infinite_intervals.
+   *
+   * @param stream_interval Callback method to process the birth and death index pairs. Has to take three arguments
+   * as input: first the dimension of the cycle, then the birth index of the cycle and third the death index of the
+   * cycle. An index always corresponds to the arrow number the event occurred (one call to @ref insert_face,
+   * @ref remove_face or @ref apply_identity is equal to one arrow and increases the arrow count by one).
+   * @param preallocationSize Maximal value among the minimum numbers of faces known to be at the same time in each
+   * complex of the filtration. It will be used to optimize the memory allocation.
+   * Default value: 0.
+   */
+  Zigzag_persistence(std::function<void(dimension_type, index, index)> stream_interval,
+                     unsigned int preallocationSize = 0)
+      : matrix_(
+            preallocationSize,
+            [this](matrix_index columnIndex1, matrix_index columnIndex2) -> bool {
+              if (matrix_.get_column(columnIndex1).is_paired()) {
+                return matrix_.get_pivot(columnIndex1) < matrix_.get_pivot(columnIndex2);
+              }
+              return birthOrdering_.birth_order(births_.at(columnIndex1), births_.at(columnIndex2));
+            },
+            [this](matrix_index columnIndex1, matrix_index columnIndex2) -> bool { return false; }),
+        numArrow_(-1),
+        stream_interval_(std::move(stream_interval)) {}
+  /**
+   * @brief Updates the zigzag persistence diagram after the insertion of the given face.
+   *
+   * @tparam BoundaryRange Range type needing size, begin and end members.
+   * @param boundary Boundary of the inserted face. The boundary should be represented by all the faces with
+   * non-zero coefficients generating it. A face should be represented by the arrow number when the face appeared for
+   * the first time in the filtration (if a face was inserted and then removed and reinserted etc., only the last
+   * insertion counts). The face range should be ordered by increasing arrow numbers.
+   * @param dimension Dimension of the inserted face.
```

Comment:
From what I can see, the only thing the dimension is used for is the output, and we wouldn't lose anything if we removed the dimension completely from this class, wrappers could handle it on their own.
I am not asking to do that (I haven't thought about it much), I just wanted to write this down.

Reply:
Yes, it is the same than for the matrix module: the dimension is only useful for the persistence diagram. I could make it optional for both.

Reply:
The tests didn't notice the problem because... there were no infinite bars? Or they were all of the same dimension?

Reply:
No, because all faces were simplices and then the matrix just deduces the dimension if I don't give any. I should construct some examples of non-simplicial filtrations by hand to avoid such problems.

Reply:
A bit error-prone indeed... But there is worse and at least it comes in handy, as simplices are still the bigger use case. 
If the size is one, the dimension will just be put to 0. Everything else will work completely alright. It is just that when returning the barcode, the dimension will be completely wrong.

Reply:
Ah, I am not completely right: there is one point where a wrong dimension will make problems and that is in the boundary matrix (with only R and not RU). The reduction algorithm uses the clearing optimization which uses the dimensions.

Reply:
> If the size is 1, the dimension will just be put to 0.

Doesn't that deserve an assert?

Reply:
> I could make it optional for both.

How horrible would it look?

Reply:
> How horrible would it look?

It won't be more horrible than any other thing in the matrix module, but you will not like it for sure. I would need to make two interval types (are make the interval inherit its dimension) and then add some `std::conditional` and `if constexpr` here and there.